lib/primitives/mtx.js

```diff
@@ -1881,6 +1890,132 @@ class CoinSelector {
+    let size = await this.tx.estimateSize();
+    const fee = this.getFee(size);
+    // add fee to target value.
+    target += fee;
+
+    // begin selection.
+
+    // while target is greater than
+    // the largest coin we have, we
+    // will keep selecting the largest coin
+
+    this.index = this.coins.length - 1;
+    let coin = this.coins[this.index];
+    let value = Amount.value(coin.value);
+    let spendingSize = await coin.estimateSpendingSize(this.getAccount);
+    let effectiveValue = value - this.getFee(spendingSize);
+    while (target >= effectiveValue) {
+      this.tx.addCoin(coin);
+      this.chosen.push(coin);
+
+      target -= effectiveValue;
+
+      this.index--;
+      if (this.index < 0)
+        return;
+
+      coin = this.coins[this.index];
+      value = Amount.value(coin.value);
+      spendingSize = await coin.estimateSpendingSize(this.getAccount);
+      effectiveValue = value - this.getFee(spendingSize);
+    }
+
+    if (target > 0) {
+      // now we are sure that target < largest unselected coin
+      // we will perform Binary search to find the smallest coin
+      // which is greater than target value
+
+      const closestCoin = await this.findSmallestGreater(target);
+      this.tx.addCoin(closestCoin);
+      this.chosen.push(closestCoin);
+    }
+
+    // update fee
+    for (const coin of this.chosen) {
+      size += await coin.estimateSpendingSize(this.getAccount);
+    }
+    this.fee = this.getFee(size);
+  }
+
+  /**
+   * Find smallest coin greater than
+   * the target using binary search
+   * @param {Number} target
+   * @returns {Coin}
+   */
+
+  async findSmallestGreater(target) {
+    // Corner cases
+    if (target <= Amount.value(this.coins[0].value))
+      return this.coins[0];
+    if (target >= Amount.value(this.coins[this.index].value))
+      return this.coins[this.index];
+
+    // begin binary search
+    let i = 0;
+    let j = this.index + 1;
+    let mid = 0;
+    while (i < j) {
+      mid = Math.floor((i + j) / 2);
+
+      // calculate effective value of coin at mid
+      const coin = this.coins[mid];
+      let size = await coin.estimateSpendingSize(this.getAccount);
+      let value = Amount.value(coin.value);
+      const effectiveValue = value - this.getFee(size);
+
+      if (effectiveValue === target)
+        return coin;
+
+      // if target is less than coin at mid
+      // then search in left part of array
+      if (target < effectiveValue) {
+        if (mid > 0) {
+          // calculate effective value of coin at mid - 1
+          const prev = this.coins[mid - 1];
+          size = await prev.estimateSpendingSize(this.getAccount);
+          value = Amount.value(prev.value);
+          const prevEffectiveValue = value - this.getFee(size);
+          // if target is greater than previous
+          // effective value, then return mid
+          if (target > prevEffectiveValue)
+            return coin;
```

Comment:
Does this guarantee that by adding only one more small coin we will be done? I'm going to modify your test and come up with some examples...